ksqldb-execution/src/main/java/io/confluent/ksql/execution/util/FunctionArgumentsUtil.java

```diff
@@ -0,0 +1,227 @@
+/*
+ * Copyright 2021 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.execution.util;
+
+import io.confluent.ksql.execution.codegen.TypeContext;
+import io.confluent.ksql.execution.expression.tree.Expression;
+import io.confluent.ksql.execution.expression.tree.FunctionCall;
+import io.confluent.ksql.execution.expression.tree.LambdaFunctionCall;
+import io.confluent.ksql.function.GenericsUtil;
+import io.confluent.ksql.function.KsqlScalarFunction;
+import io.confluent.ksql.function.UdfFactory;
+import io.confluent.ksql.function.types.GenericType;
+import io.confluent.ksql.function.types.LambdaType;
+import io.confluent.ksql.function.types.ParamType;
+import io.confluent.ksql.schema.ksql.SchemaConverters;
+import io.confluent.ksql.schema.ksql.SqlArgument;
+import io.confluent.ksql.schema.ksql.types.SqlLambda;
+import io.confluent.ksql.schema.ksql.types.SqlLambdaResolved;
+import io.confluent.ksql.schema.ksql.types.SqlType;
+import io.confluent.ksql.util.KsqlException;
+import io.confluent.ksql.util.Pair;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+public final class FunctionArgumentsUtil {
+
+  private FunctionArgumentsUtil() {
+
+  }
+
+  public static class FunctionArgumentsAndContext {
+    private final List<TypeContext> contexts;
+    private final List<SqlArgument> arguments;
+    private final SqlType returnType;
+    private final KsqlScalarFunction function;
+
+    public FunctionArgumentsAndContext(
+        final List<TypeContext> contexts,
```

Comment:
Overall I think this makes sense. We do need to track inputs somewhere though because we depend on input type ordering when creating a `SqlLambdaResolved`, so we can't just add the map keys as the input types since they don't have guaranteed ordering. I went ahead and added a list to `mapLambdaParametersToTypes` that just adds the inputs' sqlTypes, but do you think the list should be in the context instead? It doesn't need to be shared but it does feel kinda repetitive to build a map and a list of the values in the map at the same time

Reply:
I had similar thoughts when reviewing another PR by @lct45, and she convinced me that we indeed need both the list and the map because the ordering is required. Now with two passes, my understanding is that the second pass would still rely on the ordering so `SqlLambdaResolved` itself still need both list and map? maybe @lct45 can correct me here.

Reply:
Yeah that exactly right, we do need some way of keeping the inputs in the correct order so having a map alone won't do. _where_ we put the list is somewhat flexible though

Reply:
We need the order at this level (the enclosing function call). But when we recurse down to the lambda expression itself, why do we need it any more? We should know the types for all the variables.

In today's implementation we basically do:
- build up a list of argument types
- recurse down to LambdaFunction
- correlate the list with the parameter list in LambdaFunction
- clear the list (at this point the list is gone so we don't care about the order anymore)

I'm proposing we just do all 4 in 1 place and store it as a map

Reply:
Never mind - I see what you mean now. I think we can just keep the list to within the implementation of `getLambdaContextAndType` and maybe `mapLambdaParametersToTypes` (you could just build it from the mapping and the params in the expression) rather than passing it around.